src/pyrokinetics/diagnostics/__init__.py

```diff
@@ -248,3 +251,195 @@ def _invfft(f, x, y, kx, ky):
+            "gds2": gds2,
+            "gbdrift": gbdrift,
+            "gbdrift0": gbdrift0,
+            "grho": grho,
+            "gds21": gds21,
+            "gds22": gds22,
+            "rplot": R,
+            "zplot": Z,
+            "rprime": dR_dr,
+            "zprime": dZ_dr,
+            "aplot": S,
+            "aprime": dS_dr,
+        }
+
+    def ideal_ballooning_solver(self):
+        r"""
+        Adapted from ideal-ballooning-solver
+        https://github.com/rahulgaur104/ideal-ballooning-solver/blob/master/ball_scan.py
+
+        """
+
+        geometry_terms = self.gs2_geometry_terms()
+
+        dpdrho = geometry_terms["dpdrho"]
+        theta = geometry_terms["theta"]
+        bmag = geometry_terms["bmag"]
+        gradpar = geometry_terms["gradpar"]
+        cvdrift = geometry_terms["cvdrift"]
+        gds2 = geometry_terms["gds2"]
+
+        gamma, X_arr, dX_arr, g_arr, c_arr, f_arr = gamma_ball_full(
+            dpdrho, theta, bmag, gradpar, cvdrift, gds2
+        )
+
+        return gamma
+
+
+def gamma_ball_full(
+    dPdrho, theta_PEST, B, gradpar, cvdrift, gds2, vguess=None, sigma0=0.42
+):
+    # Inputs  : geometric coefficients(normalized with a_N, and B_N)
+    #           on an equispaced theta_PEST grid
+    # Outputs : maximum ballooning growth rate gamma
+    theta_ball = theta_PEST
+    # ntheta = len(theta_ball)
+
+    # Note that gds2 is (dpsidrho*|grad alpha|/(a_N*B_N))**2.
+    g = np.abs(gradpar) * gds2 / (B)
+    c = -1 * dPdrho * cvdrift * 1 / (np.abs(gradpar) * B)
+    f = gds2 / B**2 * 1 / (np.abs(gradpar) * B)
+
+    len1 = len(g)
+
+    # Uniform half theta ball
+    theta_ball_u = np.linspace(theta_ball[0], theta_ball[-1], len1)
+
+    g_u = np.interp(theta_ball_u, theta_ball, g)
+    c_u = np.interp(theta_ball_u, theta_ball, c)
+    f_u = np.interp(theta_ball_u, theta_ball, f)
+
+    # uniform theta_ball on half points with half the size, i.e., only from [0, (2*nperiod-1)*np.pi]
+    theta_ball_u_half = (theta_ball_u[:-1] + theta_ball_u[1:]) / 2
+    h = np.diff(theta_ball_u_half)[2]
+    g_u_half = np.interp(theta_ball_u_half, theta_ball, g)
+    g_u1 = g_u[:]
+    c_u1 = c_u[:]
+    f_u1 = f_u[:]
+
+    len2 = int(len1) - 2
+    A = np.zeros((len2, len2))
+
+    A = (
+        np.diag(g_u_half[1:-1] / f_u1[2:-1] * 1 / h**2, -1)
+        + np.diag(
+            -(g_u_half[1:] + g_u_half[:-1]) / f_u1[1:-1] * 1 / h**2
+            + c_u1[1:-1] / f_u1[1:-1],
+            0,
+        )
+        + np.diag(g_u_half[1:-1] / f_u1[1:-2] * 1 / h**2, 1)
+    )
+
+    # Method without M is approx 3 X faster with Arnoldi iteration
+    # Perhaps, we should try dstemr as suggested by Max Ruth. However, I doubt if
+    # that will give us a significant speedup
+    w, v = eigs(A, 1, sigma=sigma0, v0=vguess, tol=5.0e-7, OPpart="r")
+    # w, v  = eigs(A, 1, sigma=1.0, tol=1E-6, OPpart='r')
+    # w, v  = eigs(A, 1, sigma=1.0, tol=1E-6, OPpart='r')
+
+    # ## Richardson extrapolation
```

Comment:
Please replace Richardson extrapolation with Variational refinement. I'll do the same in my repo.